src/Conversion/ONNXToKrnl/Tensor/Slice.cpp

```diff
@@ -0,0 +1,98 @@
+//===----------------Slice.cpp - Lowering Slice Op----------------------=== //
+//
+// Copyright 2020 The IBM Research Authors.
+//
+// =============================================================================
+//
+// This file lowers the ONNX Slice Operator to Krnl dialect.
+//
+//===----------------------------------------------------------------------===//
+
+#include "src/Conversion/ONNXToKrnl/ONNXToKrnlCommon.hpp"
+#include "src/Dialect/ONNX/ONNXShapeHelper.hpp"
+
+using namespace mlir;
+
+struct ONNXSliceOpLowering : public ConversionPattern {
+  ONNXSliceOpLowering(MLIRContext *ctx)
+      : ConversionPattern(mlir::ONNXSliceOp::getOperationName(), 1, ctx) {}
+
+  LogicalResult matchAndRewrite(Operation *op, ArrayRef<Value> operands,
+      ConversionPatternRewriter &rewriter) const final {
+    ONNXSliceOpAdaptor operandAdaptor(operands);
+    ONNXSliceOp sliceOp = llvm::cast<ONNXSliceOp>(op);
+    auto loc = op->getLoc();
+
+    ONNXSliceOpShapeHelper shapeHelper(&sliceOp, &rewriter);
+    if (failed(shapeHelper.Compute(operandAdaptor)))
+      return op->emitError("Failed to scan Silce parameters successfully");
+
+    auto outputMemRefType = convertToMemRefType(*op->result_type_begin());
+    int64_t outputRank = outputMemRefType.getShape().size();
+    assert(outputRank == shapeHelper.outputDims.size());
+    // Insert an allocation and deallocation for the output of this operation.
+    Value alloc = insertAllocAndDeallocSimple(
+        rewriter, op, outputMemRefType, loc, shapeHelper.outputDims);
+
+    BuildKrnlLoop outputLoops(rewriter, loc, outputRank);
+    outputLoops.createDefineOp();
+    for (int ii = 0; ii < outputRank; ++ii)
+      outputLoops.pushBounds(
+          shapeHelper.context, 0, shapeHelper.outputDims[ii]);
+    outputLoops.createIterateOp();
+    rewriter.setInsertionPointToStart(outputLoops.getIterateBlock());
+
+    IndexExprContext childContext(shapeHelper.context);
+#if 0
+    printf("\nalex, make a test for affine\n");
+    Value ind0 = outputLoops.getInductionVar(0);
+    IndexExpr i1 = outerloopContex.CreateSymbolIndex(ind0);
+    IndexExpr t1;
+    t1.Sub(i1, i1);
+    t1.debugPrint("index loop i -i");
+#endif
+
+    // Proceed with the load data["i * step + start} for all dim].
+    Value loadVal;
+    SmallVector<Value, 4> loadIndices;
+    bool loadIsAffine = true;
+    for (int ii = 0; ii < outputRank; ++ii) {
+      Value loopVal = outputLoops.getInductionVar(ii);
```

Comment:
Does user need to know the order of Dims used in Helper.cpp? Or is there a default order?